(Mostly) Enable compilation with macOS arm64 emulated x64

thewilsonator opened the pull request:
as is this will still trip ` static assert(false, "longdouble_soft not supported on this platform");` because neither of `D_InlineAsm_X86_64` or` D_InlineAsm_X86` are defined, but apart from that. building DMD with LDC on the M1 succeeds.

This is getting rather annoying to deal commit juggling with so I'm pushing these changes upstream.

cc @jacob-carlborg

thewilsonator commented:
```
undefined identifier `__c_longlong`
undefined identifier `__c_ulonglong`
```
does windows stdc.config not define those?

thewilsonator commented on src/dmd/backend/bcomplex.d:
```diff
@@ -11,7 +11,7 @@ module dmd.backend.bcomplex;
 public import dmd.root.longdouble : targ_ldouble = longdouble;
 import core.stdc.math : fabs, fabsl, sqrt;
-version(CRuntime_Microsoft)
+static if (real.sizeof == 8)
```
also WTF!

This condition is true (verified with `pragma(msg)`on apple arm64 and fails as is. 
If I remove the line and unconditionally `import dmd.root.longdouble` then it works!??!?!

ibuclaw replied:
Version conditions are expanded first, then static ifs later in semantic.  You're probably unwittingly making another version statement depend on this static if, which is bad.

thewilsonator replied:
Hmm, let's just see if an unconditional import works for all systems.

jacob-carlborg commented:
Sorry, I'm not following exactly what you're doing. Are compiling DMD as an ARM64 or x86-64 binary? Which architecture is the host compiler?

thewilsonator commented:
DMD as an ARM64 binary.  Host binary is an arm ldmd2. I haven't figured out how to get it to compile as a x64 binary.

jacob-carlborg commented:
Is DMD cross-compiling to x86-64?

jacob-carlborg commented:
`longdouble.d` is used at runtime so it's fine that there's x86 inline assembly in that file?

thewilsonator commented:
> Is DMD cross-compiling to x86-64?
> longdouble.d is used at runtime so it's fine that there's x86 inline assembly in that file?

as I said this is so I can _compile_ DMD without juggling commits. I'm not running DMD.

jacob-carlborg commented:
Aha, I see.

thewilsonator committed:
(Mostly) Enable compilation with macOS arm64 emulated x64

As is this will still trip ` static assert(false, "longdouble_soft not supported on this platform");` because neither of D_InlineAsm_X86_64 or D_InlineAsm_X86 are defined but apart from that building DMD with LDC on the M1 succedes.

This makes developing DMD much less burdensome.

thewilsonator committed:
fix

thewilsonator committed:
debug windows

thewilsonator committed:
test

thewilsonator committed:
more prodding in the dark
